tensorflow-ops/src/TensorFlow/EmbeddingOps.hs

```diff
@@ -56,21 +56,33 @@ embeddingLookup :: forall a b v .
                 -> Tensor Value b
                 -- ^ A `Tensor` with type `int32` or `int64`
                 -- containing the ids to be looked up in `params`.
-                -- The ids are required to be flat on entry and have
-                -- fewer than 2^31 entries.
+                -- The ids are required to have fewer than 2^31
+                -- entries.
                 -> Build (Tensor Value a)
                 -- ^ A dense tensor with shape `shape(ids) + shape(params)[1:]`.
-embeddingLookup params ids =
-    CoreOps.dynamicStitch pindices <$> partitionedResult
-  where np = genericLength params
-        pAssignments = CoreOps.cast (ids `CoreOps.mod` np)
-        newIds = ids `CoreOps.div` np
-        originalIndices = CoreOps.range 0 (CoreOps.size ids) 1
-        -- Partition list of ids based on assignments into np separate lists
-        gatherIds = CoreOps.dynamicPartition np newIds pAssignments
-        -- Similarly, partition the original indices.
-        pindices = CoreOps.dynamicPartition np originalIndices pAssignments
-        -- Do np separate lookups, finding embeddings for plist[p] in params[p]
-        partitionedResult = zipWithM
-                            (\p g -> colocateWith p $ render $ CoreOps.gather p g)
-                            params gatherIds
+embeddingLookup [p0] ids = colocateWith p0 (render $ CoreOps.gather p0 ids)
+embeddingLookup params@(p0 : _) ids = do
+    -- Do np separate lookups, finding embeddings for plist[p] in params[p]
+    partitionedResult <- zipWithM
+                        (\p g -> colocateWith p $ render $ CoreOps.gather p g)
+                        params gatherIds
+    let unshapedResult = CoreOps.dynamicStitch pindices partitionedResult
+    -- Shape restoration is not as optimal as it would be with client
+    -- side shape tracking.
+    paramShape <- colocateWith p0 (render (shape p0))
+    let finalShape = CoreOps.concat 0 [shape ids, tailShape]
+        tailShape = CoreOps.slice paramShape (singleton 1) (singleton (-1))
+    render $ CoreOps.reshape unshapedResult finalShape
+  where
+    np = genericLength params
```

Comment:
Are you sure? There's no constraint on `a`, which is the Tensor, so the `0` is bound by the `Num` constraint on `i`, no?

(furthermore, even a `Num a` requirement on genericLength doesn't work, because it can't guarantee that the `v`  on params :: Tensor v a is  `Value` ...)